File: C#/46. Permutations.cs
/*
 * Link: https://leetcode.com/problems/permutations/
 * Author: Dung Nguyen Tien (Chris)
*/

public class Solution {
    public IList<IList<int>> Permute(int[] nums) {
        var res = new List<IList<int>>();
        
        backtrack(res, new List<int>(), nums, new HashSet<int>());
        
        return res;
    }
    
    public void backtrack(List<IList<int>> res, List<int> current, int[] nums, HashSet<int> visited) {
        if (current.Count == nums.Length) {
            res.Add(new List<int>(current));
            
            return;
        }
        
        for (var i = 0; i < nums.Length; i++) {
            if (!visited.Contains(i)) {
                visited.Add(i);
                current.Add(nums[i]);
                
                backtrack(res, current, nums, visited);
                
                current.RemoveAt(current.Count - 1);
                visited.Remove(i);
            }
        }
    }
}

File: Javascript/46. Permutations.js
/**
 * Link: https://leetcode.com/problems/permutations/submissions/
 * Author: Dung Nguyen Tien
 * @param {number[]} nums
 * @return {number[][]}
 */

var permute = function(nums) {
    // idea:
    // take the first element, permute the right = permutedRights
    // with each item of permutedRights add the first element in
    // Forexample : "abc"
    // first: a, permutedRights = [[b,c], [c,b]]
    // with each element, let say [b,c]: add a in
    // we have: [a,b,c], [b,a,c], [b,c,a]
    if (nums.length == 0) return [[]];
    
    var res = [];
    var first = nums[0];
    var rights = nums.slice(1);
    
    var permutedRights = permute(rights);
    for (let p of permutedRights) {
        for (let i = 0; i <= p.length; i++) {
            let tmp = p.slice(0);
            tmp.splice(i, 0, first);
            res.push(tmp);
        }
    }
    
    return res;
};